packer/provisioner.go

```diff
@@ -145,15 +145,67 @@ func (p *PausedProvisioner) Prepare(raws ...interface{}) error {
 
 func (p *PausedProvisioner) Provision(ctx context.Context, ui packersdk.Ui, comm packersdk.Communicator, generatedData map[string]interface{}) error {
 
-	// Use a select to determine if we get cancelled during the wait
 	ui.Say(fmt.Sprintf("Pausing %s before the next provisioner...", p.PauseBefore))
+
+	var err error
+	switch ui.(type) {
+	case *MachineReadableUi:
+		err = p.pausingNoUpdate(ctx)
+	default:
+		minimumTimeForPauseUpdate := float64(10)
+		if p.PauseBefore.Seconds() < minimumTimeForPauseUpdate {
+			err = p.pausingNoUpdate(ctx)
+		} else {
+			err = p.pausingWithUpdates(ctx, ui)
+		}
+	}
+	if err != nil {
+		return err
+	}
+
+	return p.Provisioner.Provision(ctx, ui, comm, generatedData)
+}
+
+func (p *PausedProvisioner) pausingNoUpdate(ctx context.Context) error {
 	select {
 	case <-time.After(p.PauseBefore):
 	case <-ctx.Done():
 		return ctx.Err()
 	}
+	return nil
+}
 
-	return p.Provisioner.Provision(ctx, ui, comm, generatedData)
+func (p *PausedProvisioner) pausingWithUpdates(ctx context.Context, ui packersdk.Ui) error {
+	updateTime := 10
+	ticker := time.NewTicker(time.Duration(updateTime) * time.Second)
+	defer ticker.Stop()
+	totalTime := p.PauseBefore.Seconds()
+
+	pausingCtx, cancel := context.WithTimeout(context.Background(), p.PauseBefore)
+	defer cancel()
+
+	var err error
+	for {
+		select {
+		case <-pausingCtx.Done():
+			err = pausingCtx.Err()
+			if err != nil && isTimeRemaining(totalTime) {
+				return err
+			}
+			return nil
+		case <-ticker.C:
+			totalTime -= float64(updateTime)
+			_, ok := pausingCtx.Deadline()
+			if ok && isTimeRemaining(totalTime) {
+				ui.Say(fmt.Sprintf("%v seconds left until the next provisioner", totalTime))
+			}
```

Comment:
I think here; instead of manually printing the time in seconds; we should use durafmt, like for a build here: https://github.com/hashicorp/packer/blob/96f29867ce738facd20aaf0c574c10b83a4e5c65/command/build.go#L295

This shows the first two most relevant units of duration from the Deadline. ( you could use `timeLeft := time.Until(deadline)` )

Reply:
@teddylear, sorry about the long time to review ! Do you have thought on this suggestion here ?

Reply:
Sorry I missed this! I added this to print the time remaining.